packages/teams-js/test/public/mail.spec.ts

```diff
@@ -0,0 +1,325 @@
+import { GlobalVars } from '../../src/internal/globalVars';
+import { app } from '../../src/public/app';
+import { mail } from '../../src/public/mail';
+import { Utils } from '../utils';
+
+describe('mail', () => {
+  // Use to send a mock message from the app.
+  const utils = new Utils();
+
+  beforeEach(() => {
+    utils.processMessage = null;
+    utils.messages = [];
+    utils.childMessages = [];
+    utils.childWindow.closed = false;
+    GlobalVars.frameContext = undefined;
+
+    // Set a mock window for testing
+    app._initialize(utils.mockWindow);
+  });
+
+  afterEach(() => {
+    // Reset the object since it's a singleton
+    if (app._uninitialize) {
+      app._uninitialize();
+    }
+  });
+
+  describe('openMailItem', () => {
+    const openMailItemParams: mail.OpenMailItemParams = {
+      itemId: '',
+    };
+
+    it('should not allow calls before initialization', async () => {
+      await mail
+        .openMailItem(openMailItemParams)
+        .catch(e => expect(e).toMatchObject(new Error('The library has not yet been initialized')));
+    });
+
+    it('should not allow calls from settings context', async () => {
+      await utils.initializeWithContext('settings');
+
+      await mail
+        .openMailItem(openMailItemParams)
+        .catch(e =>
+          expect(e).toMatchObject(
+            new Error('This call is only allowed in following contexts: ["content"]. Current context: "settings".'),
+          ),
```

Comment:
I know that this is not a new pattern that you're implementing here but I've always thought it was kind of weird that we matched against the entire string. E.g. if we update the mail capability to be allowed in a new context later, we have to come back and update these tests. It seems like the only thing we should care about matching against is that it is throwing an error that says "settings" is an invalid context. But I don't really know if you can practically do that. In theory it seems like we should be throwing a more specific error like `InvalidContextError` and just checking for that error and ignoring the string. But I'm sure that's a much bigger change.

You can ignore me I guess. I mean more than usual.

Reply:
I agree with everything you wrote here, including the part about it being too big of a change for this. I think in general (in the future) we're going to want to be more deliberate about how we're handling errors and error types. The functions being tested here are occasionally throwing strings instead of Errors which is not great (as an example).